src/admin_access_token.py

```diff
@@ -41,20 +42,36 @@ def __init__(self, app: ops.Application, model: ops.Model):
         self._model = model
 
     def get(self, container: ops.Container) -> typing.Optional[str]:
-        """Get admin access token.
+        """Get an admin access token.
+
+        If the admin token is not valid or it does not exist it creates one.
 
         Args:
             container: Workload container.
 
         Returns:
             admin access token or None if fails.
         """
+        admin_access_token = self._get_from_peer_relation()
```

Comment:
Shouldnt catch PeerRelationNotFound?

Reply:
Thanks for bringing this back to the foreground.

I left the exception uncaught as I thought that was a situation that is clearly something very wrong (not having a peer relation).

Reviewing the documentation peer relations events are part of the [setup phase](https://juju.is/docs/sdk/charm-lifecycle#heading--notes-on-the-setup-phase), and occur only after `install`. It also says:
```
The reason for this is that a peer relation cannot be created or destroyed ‘manually’ at arbitrary times, they either exist or not, and if they do exist, then we know it from the start.
```

Checking manually, I see that in the `install` event we do not have the peer relation (or at least it is None using `charm.model.get_relation("synapse-peers")`). It is in all other events in the setup phase, and it disappears in the teardown phase, after the `synapse_peers_relation_broken` event.

As we are using the collect status ops event to invoke logic in the container, that means that we may in effect setting the unit to error in at least the teardown phase (haven't tested it).

As a precaution, I am going to change it to return None instead of raising/catching that exception. This was the same logic that was before the PR. I will add a note to the story about the collect status to review these cases in that story.

Reply:
Good investigation, thanks.